core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3956,11 +3961,482 @@ class KafkaApis(val requestChannel: RequestChannel,
+    var shareFetchResponse : ShareFetchResponse = null
+    // Variable to store any error thrown while the handling piggybacked acknowledgements
+    var acknowledgeError : Errors = Errors.NONE
+    // Variable to store the topic partition wise result of piggybacked acknowledgements
+    var acknowledgeResult = mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]()
+
+    def isInvalidShareFetchRequest() : Boolean = {
+      // The Initial Share Fetch Request should not Acknowledge any data
+      if (shareSessionEpoch == ShareFetchMetadata.INITIAL_EPOCH && isAcknowledgeDataPresent) {
+        return true
+      }
+      false
+    }
+
+    val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
+    var shareFetchContext : ShareFetchContext = null
+    try {
+      // Creating the shareFetchContext for Share Session Handling
+      shareFetchContext = sharePartitionManager.newContext(groupId, shareFetchData, forgottenTopics, newReqMetadata)
+    } catch {
+      case e: Exception => shareFetchResponse = shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, e) match {
+        case response: ShareFetchResponse => response
+        case _ => null
+      }
+    }
+    // This check is done to make sure that there was no Share Session related error while creating shareFetchContext
+    if(shareFetchResponse == null) {
+      // check if the Request is Invalid
+      if(isInvalidShareFetchRequest()) {
+        shareFetchResponse = shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception) match {
+          case response: ShareFetchResponse => response
+          case _ => null
+        }
+      } else {
+        val erroneousAndValidPartitionData : ErroneousAndValidPartitionData = shareFetchContext.getErroneousAndValidTopicIdPartitions
+        val topicIdPartitionSeq : mutable.Set[TopicIdPartition] = mutable.Set()
+        erroneousAndValidPartitionData.erroneous.forEach {
+          case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+        }
+        erroneousAndValidPartitionData.validTopicIdPartitions.forEach {
+          case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+        }
+        shareFetchData.forEach {
+          case(tp, _) => if (!topicIdPartitionSeq.contains(tp)) topicIdPartitionSeq += tp
+        }
+
+        val authorizedTopics = authHelper.filterByAuthorized(
+          request.context,
+          READ,
+          TOPIC,
+          topicIdPartitionSeq
+        )(_.topicPartition.topic)
+
+        // Handling the Acknowledgements from the ShareFetchRequest
+        // If this check is true, then we are sure that this is not an Initial ShareFetch Request, otherwise the request would have been invalid
+        if(isAcknowledgeDataPresent) {
+          if (!authHelper.authorize(request.context, READ, GROUP, groupId)) {
```

Comment:
Query: Why do we only authorize in subsequent request i.e. when acknowledge data is present?

Reply:
Hi, thanks for the review. According to my knowledge, this auth check is only required for acknowledgement and not for fetching. So, if there's nothing to acknowledge, there won't be any need for this auth check.

Reply:
I find following line from the KIP which says following, hence shouldn't we authorize for READ on fetch as well?

```
Operations which change information about a share group (such as consuming a record) need permission to perform the READ action on the named group resource

Reply:
Also, do we need to re-authorize on every request in session? What behaviour we have on regular fetch? Does it authorize everytime or once in session?

If we do it only at session establishment, I do get that if an API key is unauthorized then existing session might continue reading/acknowledging, but re-authorization is taxing as well. Hence checking what's the flow looks like on regular fetch.
cc: @AndrewJSchofield

Reply:
As per my knowledge the regular fetch does not authorize for READ operation on the named group. Only the topics from where data is to fetched, are authorized for READ operation

Reply:
Yeah but KIP-932 defines the behaviour for auth during fetch.